base/c.jl

```diff
@@ -503,3 +503,183 @@ end
 macro ccallable(rt, def)
     expand_ccallable(rt, def)
 end
+
+module CcallMacro
+
+struct NoType end
+struct CcallError{T <: AbstractString} <: Exception
+    msg::T
+end
+Base.showerror(io::IO, e::CcallError) = print(io, "CcallError: ", e.msg)
+
+getcfunc(f) = QuoteNode(f)
+getcfunc(f::Expr) = :(($(f.args[2]), $(f.args[1])))
+hashead(symbol, _) = false
+hashead(expr::Expr, head) = expr.head === head
+
+islinum(node) = node isa LineNumberNode
+
+remove_linums(e::Expr) =
+    Expr(e.head, remove_linums.(filter(!islinum, e.args))...)
+remove_linums(s) = s
+
+"""
+Determine if there are varargs and return a vector of all arguments
+based on the call signature.
+
+returns a tuple of `(arguments, varargs)`
+"""
+function getargs(call)
+    # no arguments
+    length(call.args) < 2 &&
+        return [], []
+
+    firstarg = call.args[2]
+    hashead(firstarg, :parameters) &&
+        return call.args[3:end], firstarg.args
+
+    return call.args[2:end], []
+end
+
+"""
+       mkarg(arg)
+
+`mkarg` takes an argument and returns a tuple of symbol and type. The
+argument must have a type annotation or an error will be thrown.
+"""
+function mkarg(arg)
+    !hashead(arg, :(::)) &&
+        throw(CcallError("args in @ccall need type annotations. '$arg' doesn't have one."))
+    return (arg=arg.args[1], type=arg.args[2])
+end
+
+"""
+get the type specified on the varargs. throw if there is more than one type.
+"""
+function getvarargtype(varargs)
+    vararg_types = Set(a.type for a in varargs)
+    length(vararg_types) > 1 &&
+        throw(CcallError("varargs @ccall with different argument types not yet supported"))
+    return pop!(vararg_types)
+end
+
+"""
+    parsecall(expression)
+
+`parsecall` is an implementation detail of `@ccall
+
+it takes and expression like `:(printf("%d"::Cstring, value::Cuint)::Cvoid)`
+returns: a tuple of `(function_name, return_type, arg_types, args)`
+
+The above input outputs this:
+
+    (:printf, :Cvoid, :((Cstring, Cuint)), ["%d", :value])
+
+Note that the args are in an array, not a quote block and have to be
+appended to the ccall in a separate step.
+"""
+function parsecall(expr::Expr)
+    # setup and check for errors
+    !hashead(expr, :(::)) && begin
+        throw(CcallError("@ccall needs a function signature with a return type"))
+    end
+    rettype = expr.args[2]
+
+    call = expr.args[1]
+    !hashead(call, :call) &&
+        throw(CcallError("@ccall has to take a function call"))
+
+    # get the function symbols
+    func = getcfunc(call.args[1])
+    normalargs, varargs = getargs(call)
+
+
+    # separate annotations from names
+    normalargs = mkarg.(normalargs)
+    args = Any[a.arg for a in normalargs]
+    types = Any[a.type for a in normalargs]
+    argtypes = :(())
+    argtypes.args = types
```

Comment:
`argtypes.args = types` is idiomatic. `:(())` is somewhat odd though—use `Expr(:tuple)`

Reply:
Yes, this is a leftover from "a simpler time" in the life of the Ccall macro. Fixed in https://github.com/JuliaLang/julia/pull/32748/commits/bbd6c5af06e54f25126de65d9e0cae7c0d4cd846

Reply:
Thanks @vtjnash for providing some more experienced eyes on this review. I figured I'd just dive in and try to push it along because it's a nice PR and it seemed to be languishing :-)